src/async_api/tokio/mod.rs

```diff
@@ -0,0 +1,298 @@
+};
+
+use futures::{future::join_all, Future};
+use log::warn;
+use tokio::{sync::Mutex, task::JoinHandle};
+
+use self::io_ops::SubChannel;
+use libc::{EAGAIN, EINTR, ENODEV, ENOENT};
+
+use super::{
+    active_session::{SessionConfiguration, Worker},
+    opened_session::{OpenedFlavor, OpenedSession},
+    ActiveSession, Filesystem, SessionHandle,
+};
+
+#[cfg(all(not(feature = "libfuse"), feature = "async_impl"))]
+use crate::MountOption;
+
+#[cfg(all(feature = "libfuse", feature = "async_impl"))]
+use std::ffi::OsStr;
+
+mod channel;
+mod io_ops;
+
+#[derive(Debug)]
+pub(crate) struct TokioSession {
+    pub session_configuration: Arc<Mutex<SessionConfiguration>>,
+    /// True if the filesystem is initialized (init operation done)
+    is_initialized: AtomicBool,
+    /// True if the filesystem was destroyed (destroy operation done)
+    is_destroyed: AtomicBool,
+    /// Pipes to inform all of the child channels/interested parties we are shutting down
+    pub destroy_signals: Arc<Mutex<Vec<tokio::sync::oneshot::Sender<()>>>>,
+
+    join_handles: Arc<Mutex<Vec<JoinHandle<Result<(), io::Error>>>>>,
+    driver_join: Arc<Mutex<Option<JoinHandle<Result<(), io::Error>>>>>,
+    driver_receiver: Arc<Mutex<Option<tokio::sync::oneshot::Receiver<()>>>>,
+    channel: Arc<Mutex<Option<channel::Channel>>>,
+}
+
+impl TokioSession {
+    async fn register_destroy(&self, sender: tokio::sync::oneshot::Sender<()>) {
+        let mut guard = self.destroy_signals.lock().await;
+        guard.push(sender)
+    }
+    fn new(channel: channel::Channel) -> Self {
+        Self {
+            session_configuration: Arc::new(Mutex::new(Default::default())),
+            is_initialized: AtomicBool::new(false),
+            is_destroyed: AtomicBool::new(false),
+            destroy_signals: Arc::new(Mutex::new(Vec::default())),
+            join_handles: Arc::new(Mutex::new(Vec::default())),
+            driver_join: Arc::new(Mutex::new(None)),
+            driver_receiver: Arc::new(Mutex::new(None)),
+            channel: Arc::new(Mutex::new(Some(channel))),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl SessionHandle for TokioSession {
+    async fn destroy(&self) -> () {
+        self.is_destroyed
+            .store(true, std::sync::atomic::Ordering::SeqCst);
+        let mut guard = self.destroy_signals.lock().await;
+
+        for e in guard.drain(..) {
+            if let Err(e) = e.send(()) {
+                warn!("Unable to send a shutdown signal: {:?}", e);
+            }
+        }
+
+        let mut chan = self.channel.lock().await;
+        chan.take();
+    }
+
+    async fn wait_destroy(&self) -> () {
+        let mut d = self.driver_receiver.lock().await;
+        if let Some(e) = d.take() {
+            let _ = e.await;
+        }
+        ()
+    }
+}
+
+#[async_trait::async_trait]
+impl ActiveSession for TokioSession {
+    fn destroyed(&self) -> bool {
+        self.is_destroyed.load(std::sync::atomic::Ordering::Relaxed)
```

Comment:
Destruction doesn't have a hard sequencing guarantee/ordering. Its around quick checks for the system to be shut down to control an exit. If we race with something and get past these then it should be safe/we don't rely on these for a property. (Since once this check has occurred something else could immediately race and mark it as destroyed we can't rely on it in the model here. )

Reply:
If that makes sense i can add some comments on it

Reply:
Actually, I still don't get it. `main_request_loop` looks like it will only exit if `destroyed()` returns `true`. If I understand "relaxed" there's no guarantee that we will ever observe the write though, because it does not cause synchronization: https://en.cppreference.com/w/cpp/atomic/memory_order

Also, if we don't rely on the value returned, then can we just remove this method?

Reply:
ahh, if we never observe it then... this is wrong. I had assumed the worst case was some sort of delayed observation -- will update them.